SW5 Changes

seblund opened the pull request:
The goal of this PR is to add all the Reveaal features added by SW5. This PR will serve as a general review of the code and a discussion of necessary changes.

Major code changes should probably occur in separate pull requests so we can review them better.

(Copy of PR #144 as I could not commit directly to it)

seblund commented on src/DataReader/proto_reader.rs:
```diff
@@ -0,0 +1,273 @@
+    proto_constraint: ProtoConstraint,
+    system: &TransitionSystemPtr,
+) -> Constraint {
+    fn determine_index(clock: ProtoComponentClock, system: &TransitionSystemPtr) -> usize {
+        if clock.clock_name == "0" && clock.specific_component.is_none() {
+            0
+        } else {
+            system
+                .clock_name_and_component_to_index(
+                    &clock.clock_name,
+                    &clock.specific_component.unwrap().component_name,
+                )
+                .unwrap()
+        }
+    }
```
We need to handle quotient clocks as well somehow as they cannot have a `clock.specific_component`. This probably involves changes to protobuf.

I also dislike the usage of `"0"` as a constant name for the zero clock, i would prefer an enum type e.g. just use `None` and `Some(..)` from `Option<ProtoComponentClock>` as we already receive in proto.

seblund commented on src/DataReader/parse_queries.rs:
```diff
@@ -172,6 +175,52 @@ fn build_expression_from_pair(pair: pest::iterators::Pair<Rule>) -> QueryExpress
+    // In the following line of code, we build a BoolExprssion based on the clock constraints defined for the given location.
+    // To make BoolExprssion we use the InvariantParser parser instead.
+    // Becuase clocks is defined as c1&&c2... in the InvariantParser we replace ',' to match the format e.g., e.g., "x>0,y<5" => "x>0&&y<5"
+    let invariant_version: Option<Box<BoolExpression>> = if clock_pair.as_str().trim() != "" {
+        let clock_string = clock_pair.as_str().trim().to_string().replace(',', "&&");
+        let invariant_version = parse(&clock_string).expect("");
+        Some(Box::new(invariant_version))
+    } else {
+        None
+    };
```
We should either update the invariant grammar to include diagonal constraints or use the edge grammar guard instead here. We **need** diagonal constraints here for the feature to be useful during simulation. 

We should eventually combine the invariant grammar and edge grammar: the edge guard should be the equivalent to the invariant and it would simplify our parser code

seblund reviewed (commented):
Went through the code and found more necessary changes. 

Currently working on fixing `QueryFailure` spaghetti in a new branch.

t-lohse commented on src/tests/reachability/split_component_tests.rs:
```diff
@@ -0,0 +1,106 @@
+        ];
+    "Different structures 2")]
+    fn split_component_invalid_input(path: Vec<TransitionID>) {
+        if TransitionID::split_into_component_lists(&path).is_ok() {
+            panic!("Expected error")
+        }
```
This should probably just be an assertion
```suggestion
        assert!(TransitionID::split_into_component_lists(&path).is_err(), "Expected error")
```

t-lohse reviewed (changes requested):
I have a few things I want to bring up, though they are fairly minor, I look forward to get the SW5 changes into the main branch!

In `src/System/pruning.rs`, a lot of `.unwrap()`s, maybe in the future some better error-handling should be implemented?

Some tests should probably be refactored to use `test_case`, but that's a bigger task and probably not in the scope of this PR.

I may be mistaken, but the program will crash when an assertion is failed, so the user has to restart the program after that. And if Reveaal runs on a server with limited access, a simple mistyped component name can cause a crash? If so, I think we should implement some better error-handling to convey what went wrong to the user (and maybe utilize `debug_assert` or something instead of the raw).

seblund commented on benches/flamegraph/flamegraph_profiler_unix.rs:
```diff
@@ -0,0 +1,39 @@
+        std::fs::create_dir_all(benchmark_dir).expect("Could not create benchmark directory");
+        let flamegraph_path = benchmark_dir.join("flamegraph.svg");
+        let flamegraph_file =
+            File::create(&flamegraph_path).expect("Could not create the flamegraph file");
```
I agree, I'll make a commit fixing all clippy warnings and add `-- -D warnings` to the cli :)

seblund commented on benches/reachability_bench.rs:
```diff
@@ -0,0 +1,43 @@
+pub mod flamegraph;
+use flamegraph::flamegraph_profiler::FlamegraphProfiler;
+
+static PATH: &str = "samples/json/EcdarUniversity";
```
I think we should use const in these cases, I'll make a commit

seblund commented on src/DataReader/parse_queries.rs:
```diff
@@ -13,7 +16,7 @@ pub struct QueryParser;
 
 pub fn parse_to_query(query: &str) -> Vec<Query> {
-    let queries = parse_to_expression_tree(query);
+    let queries = parse_to_expression_tree(query).unwrap();
```
Definitely should be a a result

seblund commented on src/DataTypes/statepair_list.rs:
```diff
@@ -85,6 +85,7 @@ impl PassedStateListExt for DepthFirstWaitingStateList {
     }
 }
 
+#[allow(clippy::new_without_default)]
```
Yeah we should, in general we don't use the Default trait enough

seblund commented on src/Simulation/decision.rs:
```diff
@@ -0,0 +1,90 @@
+            .filter_map(|transition| transition.use_transition_alt(&self.state))
+            .collect();
+
+        assert_ne!(next_states.len(), 0, "Decision leads to no new states");
+        assert_eq!(
+            next_states.len(),
+            1,
+            "Ambiguous decision leads to multiple new states"
+        );
```
Panics are fine but they should be results in the future

seblund commented on src/System/local_consistency.rs:
```diff
@@ -2,49 +2,47 @@ use edbm::zones::OwnedFederation;
+///Local consistency check WITH pruning.
+pub fn is_least_consistent(system: &dyn TransitionSystem) -> ConsistencyResult {
+    if system.get_initial_location().is_none() {
+        ConsistencyFailure::no_initial_state(system)
+    } else if let Some(mut state) = system.get_initial_state() {
+        let mut passed = vec![];
+        state.extrapolate_max_bounds(system);
+        consistency_least_helper(state, &mut passed, system)
+    } else {
         warn!("Empty initial state");
-        return false;
+        ConsistencyFailure::no_initial_state(system)
```
I agree with the `match` expression for clarity. The 3 cases are subtly different, there is just only one `ConsistencyFailure` variant for representing both the `get_initial_location().is_none()` and `get_initial_state().is_none()` case.

seblund commented on src/System/query_failures.rs:
```diff
@@ -0,0 +1,723 @@
+
+impl fmt::Display for ActionSet {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let singular = false; //self.actions.len() == 1;
+        let action_type = if self.is_input { "Input" } else { "Output" };
+
+        if singular {
+            write!(
+                f,
+                "{} {:?} in system '{}'",
+                action_type, self.actions, self.system
+            )
+        } else {
+            write!(
+                f,
+                "{}s {:?} in system '{}'",
+                action_type, self.actions, self.system
+            )
+        }
```
I think it was intended to pluralize input/output to inputs/outputs if there were more than one action. Not sure why it is set to false. I don't think we really need to pluralize so we'll go with your suggestion.

seblund commented on src/Simulation/decision.rs:
```diff
@@ -0,0 +1,90 @@
+
+    /// Get all possible [`Decision`]s from a [`State`]
+    pub fn get_decisions_from_state(state: State, system: &TransitionSystemPtr) -> Vec<Decision> {
+        let mut next_decisions = vec![];
+
+        for action in system.get_actions() {
+            let possible_transitions = system.next_transitions(&state.decorated_locations, &action);
+            for t in possible_transitions {
+                if let Some(decision) = Decision::from_state_transition(state.clone(), &t, &action)
+                {
+                    next_decisions.push(decision);
+                }
+            }
+        }
+
+        next_decisions
```
It ends up looking like this because of lifetimes, I'm not sure if it is more readable.
```suggestion
        system
            .get_actions()
            .iter()
            .flat_map(|action| {
                system
                    .next_transitions(&state.decorated_locations, action)
                    .iter()
                    .filter_map(|t| Decision::from_state_transition(state.clone(), t, action))
                    .collect::<Vec<_>>()
            })
            .collect()
```

t-lohse replied:
Personally, I tend to lean towards iterators over `for` loops in general, but I can see your point. I'll let you decide :)

seblund commented:
I have fixed all the comments that I have resolved. Not sure about what to do with the last three. 

I also just noticed that the CI was changed a lot in the SW5 repo, I have changed it back such that it builds release artifacts instead and only does so on pull requests to the main branch as it was before. 
I have also changed to CI to disallow clippy warnings.

t-lohse reviewed (approved):
LGTM, very nice!
